### PR TITLE
Minor improvements for Dynamic-cp

### PR DESCRIPTION
**Pull request by xiaoyao0115**
# What does this PR do ?

# Dynamic CP support for GDN and MTP, data scheduling adaptations

## Summary

1. Add Dynamic CP support for Gated Delta Net (GDN), including Triton autotune padding to avoid repeated kernel recompilation
2. Add Dynamic CP support for MTP
3. Add MLA + Dynamic CP support, including MultiLatentAttention, AbsorbedMLA, and DSv4HybridAttention paths
4. Adapt data scheduling to the new data_iterator creation policy introduced by upstream PR #3564 — middle PP stages now create their own data_iterators, so `broadcast_to_pp_group` is removed and per-stage data field stripping is added

### 1. GDN Dynamic CP support (`gated_delta_net.py`)

- All CP-related operations (`tensor_a2a_cp2hp`/`tensor_a2a_hp2cp`, `get_parameter_local_cp`, conv1d groups, QKV split dimensions, etc.) now use the dynamic CP group resolved from `packed_seq_params`
- **Triton autotune padding**: FLA's `causal_conv1d` Triton kernel includes `NB = cdiv(total_tokens, 1024)` in its autotune key. Dynamic CP causes `total_tokens` to vary per microbatch, triggering repeated Triton autotuning (hundreds of ms each). Input is now padded to `_CONV_PAD_ALIGNMENT = 4096` boundaries to collapse `NB` into far fewer buckets. Implementation: zero-pad the input tensor along the sequence dimension, adjust `cu_seqlens[-1]` to include padding, run conv1d, then strip padding from the output

### 2. MTP dynamic CP group fix (`gpt_model.py`, `multi_token_prediction.py`)

- `MultiTokenPredictionLayer._roll_input_ids_and_position_ids`: `roll_tensor` now uses `resolve_cp_group()` to resolve the dynamic CP group from `packed_seq_params`, instead of the static `self.cp_group`
- `GPTModel.forward`: `process_mtp_loss` now passes the dynamic CP group resolved from `packed_seq_params` instead of the static `self.pg_collection.cp`

### 3. Data scheduling adaptations (`data_schedule.py`, `data_schedule_utils.py`)

Upstream PR #3564 (`Fix quantize.py script and support packed sequences in pretrain_gpt.py`) changed `is_dataset_built_on_rank` and `get_batch` in `pretrain_gpt.py` so that in packed-sequence (SFT) mode, **all** PP stages (including middle ones) build datasets and receive data_iterators. Middle stages get a `PackedSeqParams` with `cu_seqlens`/`max_seqlen` directly from their own iterator. Previously only first/last PP stages had data_iterators, and middle stages relied on `broadcast_to_pp_group` to receive metadata from the first PP stage.

Adaptations:
- **Remove `broadcast_to_pp_group`**: Middle PP stages now participate in scheduling and have their own data_iterators, making the PP-group metadata broadcast unnecessary. The function (~95 lines) and its `numpy` dependency are removed
- **Per-stage data field stripping**: Before the all-to-all reroute, data fields not needed by the current PP stage are stripped from samples. First PP keeps only `tokens`/`position_ids`, last PP keeps only `labels`/`loss_mask`, MTP stages keep all four. Middle stages only carry metadata (`cu_seqlens`, `max_seqlen`, etc.) and do not participate in data tensor all-to-all

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/ssm/gated_delta_net.py [hunk 1]
```diff
@@ -422,36 +445,44 @@ def forward(
...
         else:
             assert self.activation in ["silu", "swish"]
+            _orig_seq = qkv.shape[1]
+            _pad_n = (-_orig_seq % _CONV_PAD_ALIGNMENT)
+            _conv_input = qkv
+            _conv_cu_seqlens = cu_seqlens_q
+            if _pad_n > 0:
+                _conv_input = torch.nn.functional.pad(qkv, (0, 0, 0, _pad_n))
+                _conv_cu_seqlens = cu_seqlens_q.clone()
+                _conv_cu_seqlens[-1] += _pad_n
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/ssm/gated_delta_net.py:461]: fixed

File: megatron/core/packed_seq_params.py [hunk 2]
```diff
@@ -64,3 +64,18 @@ def __post_init__(self):
...
+) -> dist.ProcessGroup:
+    """Return the dynamic CP group from packed_seq_params when available, else the static one.
+
+    Dynamic CP assigns a per-microbatch CP group that may differ from the
+    process-group stored at model construction time.  This helper centralises
+    the resolution logic used by GPTModel, GatedDeltaNet, and MTP layers.
+    """
+    if packed_seq_params is not None and packed_seq_params.cp_group is not None:
+        return packed_seq_params.cp_group
+    return static_cp_group
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/packed_seq_params.py:80]: fixed

File: megatron/core/ssm/gated_delta_net.py [hunk 3]
```diff
@@ -137,6 +143,20 @@ def __init__(
...
+        num_key_heads_per_tp = self.num_key_heads // self.tp_size
+        num_value_heads_per_tp = self.num_value_heads // self.tp_size
+        assert num_key_heads_per_tp % self.cp_size == 0, (
+            f"GDN head-parallel CP requires cp_size ({self.cp_size}) to evenly divide "
+            f"num_key_heads per TP rank ({num_key_heads_per_tp})."
+        )
+        assert num_value_heads_per_tp % self.cp_size == 0, (
+            f"GDN head-parallel CP requires cp_size ({self.cp_size}) to evenly divide "
+            f"num_value_heads per TP rank ({num_value_heads_per_tp})."
+        )
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/ssm/gated_delta_net.py:161]: fixed

File: megatron/core/datasets/data_schedule.py [hunk 4]
```diff
@@ -550,7 +559,9 @@ def get_batch_on_this_rank_for_sequence_packing(
 
     if is_first_or_last_stage or mtp_on_this_rank:
         if is_tp_rank_0:
-            total_tokens = torch.tensor(batch['tokens'].size(0), dtype=torch.int32, device=dev)
+            # Use whichever data field is available (first stage has tokens, last has labels)
+            _data_field = batch.get('tokens', batch.get('labels'))
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule.py]: fixed

File: megatron/core/datasets/data_schedule_utils.py [hunk 5]
```diff
@@ -346,14 +250,16 @@ def create_data_iterator(
...
             if is_dynamic_cp:
                 metadata_keys.append("local_cp_size")
-            metadata = [
-                {k: sample[k] for k in metadata_keys if k in sample} for sample in new_samples
-            ]
             new_data_iterator = []
             for i in range(vpp_size):
-                if vpp_has_data is not None and vpp_has_data[i]:
+                if vpp_needs_data is not None and vpp_needs_data[i]:
                     new_data_iterator.append(RerunDataIterator(iter(new_samples)))
```
Comment (yuzhongw-nvidia) [inline megatron/core/datasets/data_schedule_utils.py]: `create_data_iterator()` gives every `vpp_needs_data=True` stage an iterator over the same `new_samples` list. But `get_batch_on_this_rank_for_sequence_packing()` mutates the returned batch dict in place: CP slicing rewrites fields, tensors are reshaped, and unused fields are set to None. 

For example, In PP=2 and VPP=2, pp0vp0 and pp0vp1 both need data. The pp0vp0 can mutate labels/loss fields before the pp0vp1 reads the same dict, causing failure or incorrect packed batch contents.

File: megatron/core/ssm/gated_delta_net.py [hunk 6]
```diff
@@ -56,6 +56,12 @@
 
 logger = logging.getLogger(__name__)
 
+# Triton's autotune key for causal_conv1d includes cdiv(total_tokens, 1024).
+# Dynamic CP causes total_tokens to vary per microbatch, triggering repeated
+# autotuning.  Aligning to this boundary collapses most variations into a
+# small number of buckets.
+_CONV_PAD_ALIGNMENT = 4096
```
Comment (yuzhongw-nvidia) [inline megatron/core/ssm/gated_delta_net.py:63]: I have two questions.

1. Why only pad for conv, but do not pad for gated delta rule?
2. Compared with temporarily padding for the conv, how about directly padding to `--max-seqlen-per-dp-cp-rank` in dataloader?

**Review by yuzhongw-nvidia — APPROVED** [review]

File: megatron/core/datasets/data_schedule.py [hunk 7]
```diff
@@ -288,24 +318,7 @@ def run(
...
-                seqlen_squared_sum_this_global_batch,
-            ) = broadcast_to_pp_group(
-                new_samples,
-                num_micro_batches,
-                seqlen_sum_this_global_batch,
-                seqlen_squared_sum_this_global_batch,
-                pp_group,
-                dev,
-                is_dynamic_cp=self.is_dynamic_cp,
-            )
```
Comment (yuzhongw-nvidia) [inline megatron/core/datasets/data_schedule.py:306]: Why remove PP broadcast here? For middle PP stages, their `new_samples` are None, and they do not have chances to get the metadata for packed seq params.

Please correct me if I misunderstand it.

**Review by yaox12 — APPROVED** [review]

File: megatron/core/datasets/data_schedule.py (hunk 7 above)
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule.py:306]: now, all pp stages would create dataloader. The docstring expired and need to be modified. I will make the changes.

File: megatron/core/ssm/gated_delta_net.py (hunk 6 above)
Comment (xiaoyao0115) [reply megatron/core/ssm/gated_delta_net.py:63]: Because only conv triggers autotune frequently, which very expensive.

File: megatron/core/datasets/data_schedule_utils.py (hunk 5 above)
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py]: Good catch, i'll fix it!
